File: src/test/java/org/bitbucket/eunjeon/seunjeon/DoubleArrayTrieTest.java
package org.bitbucket.eunjeon.seunjeon;


import org.junit.Test;

import java.util.ArrayList;
import java.util.List;

import static org.junit.Assert.assertEquals;


/**
 * Created by parallels on 7/26/15.
 */
public class DoubleArrayTrieTest {
    @Test
    public void testSearch() {
        List<String> keywords = new ArrayList<String>();
        keywords.add("가가");
        keywords.add("감자");
        keywords.add("고구마");
        keywords.add("고구마");
        keywords.add("양파");
        keywords.add("헬렌켈러");
        keywords.add("흐라");

        //int [] values = {1, 2, 3, 4, 5};

        DoubleArrayTrie da = new DoubleArrayTrie();
        da.build(keywords);
        //da.build(keywords, null, values, keywords.size());
        assertEquals(2, da.exactMatchSearch("고구마"));
        assertEquals("[6]", da.commonPrefixSearch("흐라").toString());
        assertEquals("[2]", da.commonPrefixSearch("고구마").toString());
        assertEquals("[2]", da.commonPrefixSearch("감고구마", 1, 4, 0).toString());
    }
}

File: src/test/scala/org/bitbucket/eunjeon/seunjeon/DoubleArrayTrieTest.scala
package org.bitbucket.eunjeon.seunjeon

import java.io.{FileInputStream, InputStream}

import org.scalatest._

import scala.io.Source

class DoubleArrayTrieTest extends FunSuite {

  test("build") {
    val daTrie = DoubleArrayTrieBuilder().
      add("a", 10).
      add("ab", 20).
      add("acd", 30).
      build()
    assert("10,20" == daTrie.commonPrefixSearch("ab").mkString(","))
  }

  test("write read") {
    DoubleArrayTrieBuilder().
      add("a", 10).
      add("ab", 20).
      add("abc", 30).
      add("bc", 35).
      add("가", 40).
      add("가나다", 70).
      build().write(new java.io.File("test_trie.dat"))

    val newDaTrie = DoubleArrayTrie(new java.io.File("test_trie.dat"))
    assert("10,20,30" == newDaTrie.commonPrefixSearch("abc").mkString(","))
    assert("10,20" == newDaTrie.commonPrefixSearch("ab").mkString(","))
    assert("40,70" == newDaTrie.commonPrefixSearch("가나다").mkString(","))
    assert("" == newDaTrie.commonPrefixSearch(" 가").mkString(","))
    assert("40" == newDaTrie.commonPrefixSearch("가 나다").mkString(","))
  }

  test("trie performance") {
    val cl = classOf[DoubleArrayTrieTest].getClassLoader
    val inputStream = cl.getResourceAsStream("long_sentence.txt")

    val lines = Source.fromInputStream(inputStream, "UTF-8").getLines().toSeq.filter(_.nonEmpty)
    val builder = DoubleArrayTrieBuilder()
    lines.foreach(builder.add(_, 0))
    val trie = builder.build()
    println(trie.trieInfo)

    (0 to 10000).foreach { c =>
      lines.foreach( l =>
        assert(trie.commonPrefixSearch(l).nonEmpty, s"'$l' not found")
      )
    }
  }

}
